### DISCUSSION TITLE
1.4.0

### DISCUSSION THREAD

**Discussion by Mattk70** [Mattk70/Chirpity-Electron#60] [Announcements]
A lot of work behind the scenes has eliminated the need to extract compressed files and cache the resulting WAV files in Chirpity. Now, loading an mp3, mp4 or other compressed file happens immediately, without the 'Decompressing file' stage. Using a streaming pipeline, these files are read, decompressed on the fly and rendered to the UI. This approach has also unlocked the key new feature for this release...

- **There is now an option to normailse audio levels in settings**
- Exported detection CSV files now contain latitude, longitude and the location of the detections, according to the files' recording location. (see #57 )
-  The c. 25,000 file analysis limit has been removed

#### Bug fixes
- Where files in a batch analysis were recorded concurrently, the results for each timestamp are not overwritten. (Fixes #59 )

**Full Changelog**: https://github.com/Mattk70/Chirpity-Electron/commits/v1.4.0

<hr /><em>This discussion was created from the release <a href='https://github.com/Mattk70/Chirpity-Electron/releases/tag/v1.4.0'>1.4.0</a>.</em>

**Comment by Vollpflock** [comment]
Normalizing audio is a great feature, thank you. How is this supposed to work? I opened a very quiet file and activated the "Normalize audio" function in the settings, but the audio levels have not changed.

**Reply by Mattk70** [reply]
It should change the levels. For example, first without normalisation then with normalisation enabled:

<img width="400" alt="Screenshot 2024-03-02 at 10 30 51" src="https://github.com/Mattk70/Chirpity-Electron/assets/61826357/b22ba29e-d3c6-4d7b-aeb1-2626237da41e">
<img width="400" alt="Screenshot 2024-03-02 at 10 32 52" src="https://github.com/Mattk70/Chirpity-Electron/assets/61826357/43985282-c986-4c41-a976-4107853fc59e">

**Reply by Vollpflock** [reply]
Thank you, it now works with louder files. I tried it with very quiet files and it didn't work. Don't worry, nobody has such quiet files except me :)

**Reply by Mattk70** [reply]
I suspect if the section of file you are looking at is in fact all silence, the filter will not change that.

**Reply by Vollpflock** [reply]
You're right, didn't notice that it was only at the beginning very quiet.

**Reply by Vollpflock** [reply]
I'm not sure, but often the normalization doesn't work. I tested it with a number of short files, many of which were not normalized. No error codes. Do you have any idea what could be causing this?

**Reply by Mattk70** [reply]
Here's the code:

```
        let command = ffmpeg(file)
            .seekInput(start)
            .duration(end - start)
            .format('s16le')
            .audioChannels(1) // Set to mono
            .audioFrequency(24_000) // Set sample rate to 24000 Hz (always - this is for wavesurfer)
            .output(stream, { end:true });
        if (STATE.audio.normalise) command = command.audioFilter("loudnorm=I=-16:LRA=11:TP=-1.5");
```

If it doesn't work, let's eliminate some possibilities.

1.  STATE.audio.normalise is false. It's set to true when you enable the switch in settings.
2. You have somehow got a version of ffmpeg that doesn't support the audio filters. This would produce an error, "cannot open filters" or something like that. 

If either of these is true, no file will appear normalised. So - is your experience *just* for some files and not others?

If this just applies to some files, and not all, it must be the way the filter works. It runs a single-pass dynamic normalisation, as documented here: https://ffmpeg.org/ffmpeg-filters.html#loudnorm

**Reply by Vollpflock** [reply]
Thank you. Yeah, it works for some files and not for others. I have normalization turned on and am using ffmpeg version 6.1.1, which is the latest version, if I'm not mistaken. I've also reinstalled Chirpity, but with the same support files. Maybe I should delete them first?

**Reply by Mattk70** [reply]
Sorry, what do you mean by “support files”? 

Chirpity bundles its own Ffmpeg. Only if you install the app from source might you be using a different one.

**Comment by Vollpflock** [comment]
The helper files that are found in the library. 

Yeah you're right, it's probably because I downloaded it directly from Github. However, the already analyzed files are still not normalized after reinstalling Chirpity. I need to do some more testing.

**Comment by Mattk70** [comment]
Nope. I am still in the dark as to the files you're referring to. Chirpity doesn't have a library with helper files. It has a help menu with support files, I guess. But these have no effect on the operation of the application itself.

Perhaps you could share the steps you take to reproduce this issue, and share an example of a file which is not normailised? N.B. Though you haven't mentioned exporting a file, the normalisation that happens is only visible /audible in the Chirpity UI, it is not applied to exported clips.

**Comment by Vollpflock** [comment]
You can see from my name that I'm not the smartest :) Yeah, I probably meant those. 

Is there a way I could share a sample audio file with you?

**Reply by Mattk70** [reply]
In this message box, there's a menu that allows you to attach files. WAV's aren't permitted, but you can zip up a wav file and attach that.
[292253.zip](https://github.com/Mattk70/Chirpity-Electron/files/14759503/292253.zip)

**Reply by Vollpflock** [reply]
Thanks, didn't know that zip-files work. [2024-03-10_18-32-58 copy.wav.zip](https://github.com/Mattk70/Chirpity-Electron/files/14762515/2024-03-10_18-32-58.copy.wav.zip)

P.S: I know that normalization is not applied to exported clips. Is there a specific reason for it or is it just difficult to code?

**Reply by Mattk70** [reply]
Cool. I see what you mean, most likely. The file *is* being normalised, however there is high amplitude low-frequency noise so the impact on the frequencies of interest is very slight. Essentially, a small boost to the low frequency noise gets the file to the peak amplitude.

It turns out it would be super easy to normailse audio on export.  Maybe I should, I'll have a think about that. It's a bit of a re-think because ideally, you would apply the high passs and low shelf filters _before_ normalising. The way things work at the moment, normalise is done first, with ffmpeg, and other filters are done with web audio after. The only way to sequence the filters as I think would be best is to ditch the web audio filtering and do it all in ffmpeg. This may work out fine, but it may also be a lot slower, I'll have to check it out.

**Reply by Vollpflock** [reply]
Thank you. How much are the files normalized? With Audacity, I normally do aim for -3db.

**Reply by Mattk70** [reply]
The normalisation details are here, along with a link to the docs explaining those settings: https://github.com/Mattk70/Chirpity-Electron/discussions/60#discussioncomment-8892555

**Comment by Vollpflock** [comment]
I think the easiest solution will be to normalize them before opening in Chirpity
